drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,367 @@
+		k_sem_take(&m_data.acq_sem, K_FOREVER);
+
+		/* Wait for ADC data ready */
+		while ((analog_read_reg8(AREG_ADC_DATA_STATUS) & ADC_DATA_READY)
+				!= ADC_DATA_READY) {
+		}
+
+		/* Perform read */
+		adc_code = (adc_b91_get_code() / m_data.resolution_divider);
+		if (!m_data.differential) {
+			/* Sign bit is not used in case of single-ended configuration */
+			adc_code = adc_code * 2;
+		}
+		*m_data.buffer++ = adc_code;
+
+		/* Power off ADC */
+		adc_power_off();
+
+		/* Release ADC context */
+		adc_context_on_sampling_done(&m_data.ctx, dev);
+	}
+}
+
+/* ADC Driver initialization */
+static int adc_b91_init(const struct device *dev)
+{
+	k_sem_init(&m_data.acq_sem, 0, 1);
+
+	k_thread_create(&m_data.thread, m_data.stack,
+			CONFIG_ADC_B91_ACQUISITION_THREAD_STACK_SIZE,
+			(k_thread_entry_t)adc_b91_acquisition_thread,
+			(void *)dev, NULL, NULL,
+			CONFIG_ADC_B91_ACQUISITION_THREAD_PRIO,
+			0, K_NO_WAIT);
+
+	adc_context_unlock_unconditionally(&m_data.ctx);
+
+	return 0;
+}
+
+/* API implementation: channel_setup */
+static int adc_b91_channel_setup(const struct device *dev,
+				 const struct adc_channel_cfg *channel_cfg)
+{
+	adc_ref_vol_e vref;
+	adc_pre_scale_e pre_scale;
+	adc_sample_freq_e sample_freq;
+	adc_sample_cycle_e sample_cycl;
+	adc_input_pin_def_e input_positive;
+	adc_input_pin_def_e input_negative;
+
+	/* Check channel ID */
+	if (channel_cfg->channel_id > 0) {
+		LOG_ERR("Only channel 0 is supported.");
+		return -EINVAL;
+	}
+
+	/* Check reference */
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Selected ADC reference is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check gain */
+	switch (channel_cfg->gain) {
+	case ADC_GAIN_1:
+		pre_scale = ADC_PRESCALE_1;
+		break;
+	case ADC_GAIN_1_4:
+		pre_scale = ADC_PRESCALE_1F4;
+		break;
+	default:
+		LOG_ERR("Selected ADC gain is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check acquisition time */
+	switch (channel_cfg->acquisition_time) {
+	FOR_EACH(CASE_ACK_TIME_TO_HAL_SAMPLE_CYC, (),
+		 3, 6, 9, 12, 15, 18, 21, 24, 27,
+		 30, 33, 36, 39, 42, 45, 48)
+
+	case ADC_ACQ_TIME_DEFAULT:
+		sample_cycl = ADC_SAMPLE_CYC_6;
+		break;
+
+	default:
+		LOG_ERR("Selected ADC acquisition time is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check for valid pins configuration */
+	input_positive = adc_b91_get_pin(channel_cfg->input_positive);
+	input_negative = adc_b91_get_pin(channel_cfg->input_negative);
+	if ((input_positive == (uint8_t)ADC_VBAT || input_negative == (uint8_t)ADC_VBAT) &&
+		channel_cfg->differential) {
+		LOG_ERR("VBAT pin is not available for differential mode.");
+		return -EINVAL;
+	} else if (channel_cfg->differential && (input_negative == (uint8_t)NOINPUTN)) {
+		LOG_ERR("Negative input is not selected.");
+		return -EINVAL;
+	}
+
+	/* Init ADC */
+	m_data.differential = channel_cfg->differential;
+	vref = DT_STRING_TOKEN(DT_DRV_INST(0), vref);
+	sample_freq = DT_STRING_TOKEN(DT_DRV_INST(0), sample_freq);
```

Comment:
please avoid instance based macros within drivers, better make use of driver's config for consistency with most drivers in tree

Reply:
Done.